pykafka/producer.py

```diff
@@ -395,9 +405,26 @@ def queue_reader():
                     # surface all exceptions to the main thread
                     self.producer._worker_exception = sys.exc_info()
                     break
+            try:
```

Comment:
I think I prefer the way you did this before, through `Producer._update_leaders`. Here, there's no control over when exactly these messages make it into `_waiting_messages` (they might wait for the lock for a while?), so chances are they only make it into there after `Producer._update` finishes - in which case they could be there a long time.